### ISSUE TITLE
partial sync

### ISSUE THREAD

**Issue by monperrus** [pimutils/vdirsyncer#468]
Is there a way to do unidirectional partial sync with vdirsyncer?

That is, considering to calendars A and B: 

* all events of A are put in B
* if an event is moved in A it is moved in B
* if an event is deleted in A it is deleted in B
* if an event coming from A is moved or deleted in B, an error is displayed
* (B may contain events that are not in A)

**Comment by untitaker** [comment]
Does B contain events that are not in A?

**Comment by monperrus** [comment]
> Does B contain events that are not in A?

Yes

**Comment by untitaker** [comment]
Sorry, vdirsyncer doesn't support that. The closest you get to it is by setting `read_only = true` on `A`:

*  ✔️ all events of A are put in B
* ✔️ if an event is moved in A it is moved in B
* ✔️ if an event is deleted in A it is deleted in B
* ❌if an event coming from A is moved or deleted in B, **the event is restored to the original state, the change reverted**
* ❌**events from B that are not in A are deleted**

For such purposes I usually create a new calendar to sync A into.

**Comment by untitaker** [comment]
@monperrus 

>For such purposes I usually create a new calendar to sync A into.

Does this satisfy your usecase?

**Comment by monperrus** [comment]
No. There's one thing I really need: *events from B that are not in A
must be **kept**.

*

**Comment by untitaker** [comment]
Could you elaborate on your concrete usecase?

**Comment by monperrus** [comment]
My colleague maintain a calendar, and I must merge it into one that I
maintain and publish myself.

**Comment by untitaker** [comment]
Funny, I now have a usecase for this myself :)

Proposed API:

    [pair foobar]

    # A is read-only
    a = foo

    # B is your calendar that you maintain
    b = bar

    # Current behavior: Revert all changes done on B
    change_read_only = "revert"

    # Ignore changes (perhaps print a warning though?)
    change_read_only = "skip"

    # Error on changes
    change_read_only = "error"

Probably `"error"` should become the default.

Also I wonder if it would be useful to add a config option to treat edited
events differently from deleted events. For example, the user could configure:
Edited events are reverted, deleted events are skipped. But maybe for another
time.

**Comment by untitaker** [comment]
Also this needs to happen in coordination with https://github.com/pimutils/vdirsyncer/issues/453. I am not sure if those two config changes combined would look pretty.

**Comment by untitaker** [comment]
If #453 lands it would probably be better to call the `change_read_only` option `partial_sync` instead. What do you think @monperrus?

**Comment by monperrus** [comment]
Excellent!
I'm trying `partial_sync = ignore` and `partial_sync = revert` but both delete the entries in B that have only been added to B. They should not be deleted.

What do you think?

**Comment by untitaker** [comment]
You need to set one storage to `read_only`. Please file pull requests against
the documentation in `docs/` (particularly `config.rst`) to clarify.

**Comment by monperrus** [comment]
> You need to set one storage to `read_only`

OK. Does  `read_only` implies partial_sync?

**Comment by untitaker** [comment]
Yes, by default `partial_sync = revert` though.

**Comment by untitaker** [comment]
Needs better docs. @monperrus Are you interested in writing some sort of tutorial?

**Comment by monperrus** [comment]
>  @monperrus Are you interested in writing some sort of tutorial?

yes

**Comment by monperrus** [comment]
Even with `read-only=true`, both `partial_sync = ignore` and `partial_sync = revert` but both delete the entries in B that have only been added there. Seems that there is a bug.

**Comment by untitaker** [comment]
Where did you set `read_only`?

**Comment by untitaker** [comment]
I suspect that you have set it in the pair section, where vdirsyncer ignores it. I did some changes to master such that you get an error if you add arbitrary options to the pair section.

In the example above you need to add it to `A`, as that is the storage where you have no write-access. However, it should be automatically set for the `http` storage type.

**Comment by untitaker** [comment]
Hmm.

`partial_sync = revert` is supposed to delete any added events. It "reverts" all changes that are done to B.
`partial_sync = ignore` is not supposed to do anything.
`partial_sync = error` is supposed to abort the sync.

I use `ignore` already successfully...

**Comment by monperrus** [comment]
So `ignore` is not meant to delete the newly added events in B?

Is there a test case that enforces this behavior? (because I'm pretty sure it deleted the event, let me try again)

**Comment by untitaker** [comment]
There is one, I just added another one.

**Comment by monperrus** [comment]
do you mean 2e3ebd814acfa1f4aebe8967ec5ccfb71b5fb88a?

**Comment by untitaker** [comment]
Yes, exactly. The function above too.

**Comment by monperrus** [comment]
In this test, I don't see an item added to B after first sync (only deleted events), is that correct?

**Comment by untitaker** [comment]
There is a more complex test in the second test, I've updated the first one to match your test.

**Comment by monperrus** [comment]
it works. I now do extensive testing.

I would suggest to have ignore by default instead of revert, because it's safer and more conservative.

**Comment by untitaker** [comment]
Not possible, this would change the behavior for existing setups.

On 5 October 2016 10:26:36 CEST, Martin Monperrus <notifications@github.com> wrote:
>it works. I now do extensive testing.
>
>I would suggest to have ignore by default instead of revert, because
>it's safer and more conservative.
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/issues/468#issuecomment-251613132

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by untitaker** [comment]
Also if I had the choice I'd rather have error as default.

On 5 October 2016 10:26:36 CEST, Martin Monperrus <notifications@github.com> wrote:
>it works. I now do extensive testing.
>
>I would suggest to have ignore by default instead of revert, because
>it's safer and more conservative.
>
>-- 
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/issues/468#issuecomment-251613132

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by untitaker** [comment]
I'm thinking of making that parameter mandatory when one storage is read-only.

**Comment by untitaker** [comment]
Also @monperrus are you still interested in writing a tutorial? I'd like to release the next version asap and don't want this feature left undocumented.

**Comment by untitaker** [comment]
Nevermind, I fixed it myself. I'll leave the current parameter on `revert` for backwards compat.

**Comment by monperrus** [comment]
Great, thanks! Sorry for not having found the time to work on the doc.

**Comment by Black-Fry** [comment]
I am having that very same issue - entries on b are being deleted because they do not exist on a. I am trying to sync an Outlook-calendar (a) with a Nextcloud-calendar (b). While a (type: http) shall be read-only and pass several busy/free status information on to b, b (type: caldav) shall not pass any information on to a. But b should keep information, that is not present on a.

The above discussion suggests that "partial_sync = 'ignore' " would solve the problem. Unfortunately, I receive this error message on discovery: "storage doesn't take the parameters: partial_sync" - no matter which type is used (http AND caldav).

Is the partial_sync option still available on vdirsyncer 0.18 or is there another work-around present, that I was only too blind to see?

**Comment by Black-Fry** [comment]
All fine now! Just discovered that I had a misunderstanding on reading the manual. Partial_Sync needs to be set within the "pair" part of the configuration. I placed it wrongly within the config of part of Sync-Party A. Don't forget the ".." by naming the value, e g "ignore".
